### PR TITLE
Thread restructuring

### LINKED ISSUE #1032: Move all thread code from `api` to `threads` app
The `api` app contains code that should be moved to the `threads` app. Move all of the following code to the `threads` app, fix any related imports, and make migrations (if necessary).

- [x] related code in `api/models/` to `threads/models.py`
    - [x] [Activity](https://github.com/CiviWiki/OpenCiviWiki/blob/develop/project/api/models/activity.py)
    - [x] [Civi](https://github.com/CiviWiki/OpenCiviWiki/blob/develop/project/api/models/civi.py)
    - [x] [Fact](https://github.com/CiviWiki/OpenCiviWiki/blob/develop/project/api/models/fact.py)
    - [x] [Rationale](https://github.com/CiviWiki/OpenCiviWiki/blob/develop/project/api/models/rationale.py)
    - [x] [Rebuttal](https://github.com/CiviWiki/OpenCiviWiki/blob/develop/project/api/models/rebuttal.py)
    - [x] [Response](https://github.com/CiviWiki/OpenCiviWiki/blob/develop/project/api/models/response.py)
    - [x] [Thread](https://github.com/CiviWiki/OpenCiviWiki/blob/develop/project/api/models/thread.py)
    - [x] ThreadManager
    - [x] CiviManager 
    - [x] CiviImageManager
    - [x] CiviImage
    - [x] ActivityManager
- [x] related code in `api/read.py` -> `threads/api.py`
    - [x] [`get_thread`](https://github.com/CiviWiki/OpenCiviWiki/blob/db319670f20d2e266db5814ed03f9209105365ef/project/api/read.py#L125)
    - [x] [get_civi](https://github.com/CiviWiki/OpenCiviWiki/blob/db319670f20d2e266db5814ed03f9209105365ef/project/api/read.py#L189)
    - [x] [get_civis](https://github.com/CiviWiki/OpenCiviWiki/blob/db319670f20d2e266db5814ed03f9209105365ef/project/api/read.py#L201)
    - [x] [get_responses](https://github.com/CiviWiki/OpenCiviWiki/blob/db319670f20d2e266db5814ed03f9209105365ef/project/api/read.py#L213)
- [x] related code in `api/serializers.py` -> `threads/serializers.py`
    - [x] [CiviImageSerializer](https://github.com/CiviWiki/OpenCiviWiki/blob/db319670f20d2e266db5814ed03f9209105365ef/project/api/serializers.py#L120)
    - [x] [CiviSerializer](https://github.com/CiviWiki/OpenCiviWiki/blob/db319670f20d2e266db5814ed03f9209105365ef/project/api/serializers.py#L130)
    - [x] [CiviListSerializer](https://github.com/CiviWiki/OpenCiviWiki/blob/db319670f20d2e266db5814ed03f9209105365ef/project/api/serializers.py#L182)
    - [x] [ThreadSerializer](https://github.com/CiviWiki/OpenCiviWiki/blob/db319670f20d2e266db5814ed03f9209105365ef/project/api/serializers.py#L225)
    - [x] [ThreadListSerializer](https://github.com/CiviWiki/OpenCiviWiki/blob/db319670f20d2e266db5814ed03f9209105365ef/project/api/serializers.py#L263)
    - [x] [ThreadDetailSerialzier](https://github.com/CiviWiki/OpenCiviWiki/blob/db319670f20d2e266db5814ed03f9209105365ef/project/api/serializers.py#L293)
- [x] [thread-related URL patterns](https://github.com/CiviWiki/OpenCiviWiki/blob/db319670f20d2e266db5814ed03f9209105365ef/project/api/urls.py#L9) in `api/urls.py` -> `threads/urls.py` 
- [x] related code in `api/write.py` to `threads/api.py`
    - [x] [new_thread](https://github.com/CiviWiki/OpenCiviWiki/blob/28ba0b20531575cdf5036b703d3b2ff83cbf81ad/project/api/write.py#L32) 
    - [x] [createCivi](https://github.com/CiviWiki/OpenCiviWiki/blob/28ba0b20531575cdf5036b703d3b2ff83cbf81ad/project/api/write.py#L65) - rename to `create_civi`
    - [x] [rateCivi](https://github.com/CiviWiki/OpenCiviWiki/blob/28ba0b20531575cdf5036b703d3b2ff83cbf81ad/project/api/write.py#L142) - rename to `rate_civi`
    - [x] [editCivi](https://github.com/CiviWiki/OpenCiviWiki/blob/28ba0b20531575cdf5036b703d3b2ff83cbf81ad/project/api/write.py#L189) - rename to `edit_civi`
    - [x] [deleteCivi](https://github.com/CiviWiki/OpenCiviWiki/blob/28ba0b20531575cdf5036b703d3b2ff83cbf81ad/project/api/write.py#L226) - rename to `delete_civi`
    - [x] [editThread](https://github.com/CiviWiki/OpenCiviWiki/blob/28ba0b20531575cdf5036b703d3b2ff83cbf81ad/project/api/write.py#L243) - rename to `edit_thread`
    - [x] [uploadCiviImage](https://github.com/CiviWiki/OpenCiviWiki/blob/28ba0b20531575cdf5036b703d3b2ff83cbf81ad/project/api/write.py#L396) - rename to `upload_civi_image`
    - [x] [check_image_with_pil](https://github.com/CiviWiki/OpenCiviWiki/blob/28ba0b20531575cdf5036b703d3b2ff83cbf81ad/project/api/write.py#L439)
    - [x] [uploadThreadImage](https://github.com/CiviWiki/OpenCiviWiki/blob/28ba0b20531575cdf5036b703d3b2ff83cbf81ad/project/api/write.py#L449) - rename to `upload_thread_image`
- [x] create database migrations with `python manage.py makemigrations`

### PR DESCRIPTION
**Pull request by AbhijithGanesh**
Closes #1032

### CODE REVIEW

File: project/threads/utils.py
Comment (brylie) [inline project/threads/utils.py]: This is related to `accounts` so can be deleted from your pull request.
```suggestion
```

**Review by brylie — CHANGES_REQUESTED** [review]

**Review by brylie — CHANGES_REQUESTED** [review]
Great work so far! Pretty much all we need are the serializers and database migrations :sunglasses:

**Comment by AbhijithGanesh** [comment]
Atleast then I don't need to rebase again😂. I was kind of disappointed at first but then understood the purpose of it.

**Comment by AbhijithGanesh** [comment]
@brylie please review this, excluding PEP 8 Changes, are there changes I need to make?

**Comment by brylie** [comment]
@AbhijithGanesh are you able to resolve the conflicts in this branch? If not, you can grant me write access and I can help out. I'm a bit tired tonight but might be able to help out in a day or so.

**Comment by AbhijithGanesh** [comment]
I can't see the conflicts

**Comment by brylie** [comment]
image

It shows the list of conflicts below the conversation.

**Comment by brylie** [comment]
These instructions might be helpful: 

https://docs.github.com/en/github/collaborating-with-pull-requests/addressing-merge-conflicts/resolving-a-merge-conflict-using-the-command-line

**Comment by AbhijithGanesh** [comment]
I can't resolve/ view the conflicts here because only the ones with write access can see, I shall do it on my fork though

**Comment by AbhijithGanesh** [comment]
Resolved it.

**Comment by gorkemarslan** [comment]
Because the project has constantly been changed and you are the one who is unlucky. Your task is more difficult than ours and prone to conflicting. In the end, it is inevitable to have so many conflicts.

I really understand your situation because trying, starting from the very beginning, again and again, makes you exhausted.

So congratulations and big thanks for your effort. :)

**Comment by gorkemarslan** [comment]
@AbhijithGanesh, could you take a look at the last line of the test?

`ImportError: cannot import name 'get_account' from 'threads.utils' (/home/runner/work/OpenCiviWiki/OpenCiviWiki/project/threads/utils.py)`

File: project/accounts/api.py
Comment (brylie) [inline project/accounts/api.py:21]: This should fix the error mentioned by @gorkemarslan 

```suggestion
from api.utils import get_account
from threads.utils import json_response
```

File: project/threads/permissions.py
Comment (brylie) [inline project/threads/permissions.py:2]: Related to error mentioned by @gorkemarslan 

```suggestion
from api.utils import get_account
```

**Comment by brylie** [comment]
@AbhijithGanesh perhaps the `get_account` function should be moved to `core/utils.py`. Then, the files that use `get_account` can be updated with the following import:

```py
from core.utils import get_account
```

File: project/threads/utils.py
Comment (brylie) [inline project/threads/utils.py:24]: Let's move this function to `account/utils.py` since it is related to accounts. Then, do a search for all files that use `get_account`  and add the correct import:

```py
from accounts.utils import get_account
```

**Review by brylie — CHANGES_REQUESTED** [review]

File: project/threads/utils.py
Comment (AbhijithGanesh) [reply project/threads/utils.py:24]: I will make the changes

Comment (AbhijithGanesh) [reply project/threads/utils.py:24]: @brylie Please check

File: project/accounts/api.py
Comment (gorkemarslan) [inline project/accounts/api.py]:
```suggestion
from threads.utils import json_response
from accounts.utils import get_account
```

`json_response` is in the `threads.utils` file not in the `accounts.utils`.

**Review by gorkemarslan — CHANGES_REQUESTED** [review]

File: project/core/urls.py
Comment (gorkemarslan) [inline project/core/urls.py]:
```suggestion
from core.router import CiviWikiRouter
```
You should not delete this line.

**Review by gorkemarslan — CHANGES_REQUESTED** [review]

**Comment by gorkemarslan** [comment]
@AbhijithGanesh, could you execute `python manage.py makemigrations` and `python manage.py test` before commiting. Thus, you can be sure that there is no more import error. `makemigrations` is super powerful to debug your code. 🤓

**Comment by gorkemarslan** [comment]
Your code fails because:
`django.db.migrations.exceptions.NodeNotFoundError: Migration accounts.0003_account dependencies reference nonexistent parent node ('api', '0040_delete_invitation')`

Could you consider this? I think the problem is solved after making migrations.

> @AbhijithGanesh, could you execute `python manage.py makemigrations` and `python manage.py test` before commiting. Thus, you can be sure that there is no more import error. `makemigrations` is super powerful to debug your code. 🤓

**Comment by AbhijithGanesh** [comment]
Is it possible that I clear account migrations because it has a lot of references to api

**Comment by AbhijithGanesh** [comment]
@gorkemarslan 
 I wanted to remove the pre-existing migrations because they have references to api. Would that be fine?

**Comment by gorkemarslan** [comment]
I think we shouldn't delete old migrations. Migrations also store historical records of what you've applied on your DB for whole the time. Before production, that can be discussed again but we all should be suspicious to take an action. So, except for squashing and making migrations, we shouldn't do anything about migrations. For [reference](https://stackoverflow.com/questions/28404461/can-i-delete-the-django-migration-files-inside-migrations-directory).

Notice that `api` app is not imported migrations. So deleting `api` app won't damage our project.

**Comment by AbhijithGanesh** [comment]
> 
> 
> I think we shouldn't delete old migrations. Migrations also store historical records of what you've applied on your DB for whole the time. Before production, that can be discussed again but we all should be suspicious to take an action. So, except for squashing and making migrations, we shouldn't do anything about migrations. For [reference](https://stackoverflow.com/questions/28404461/can-i-delete-the-django-migration-files-inside-migrations-directory).
> 
> Notice that `api` app is not imported migrations. So deleting `api` app won't damage our project.

Sure, I am working on the migrations keeping in mind all the points
but I'd like to politely remind everyone about the migration tree I had mentioned in one of the earlier issues #1017

**Comment by AbhijithGanesh** [comment]
@gorkemarslan and @brylie I fixed the migrations and the PEP8 errors too. Can this be merged.

**Comment by gorkemarslan** [comment]
Tests are successful and you've already spent too much effort. I think we should approve that PR. After merging, it would be easier to detect what can be done. Then, one can create a new PR to fix them. 
I try to follow but I've lost in changes. 🤕

**Comment by AbhijithGanesh** [comment]
> 
> 
> Tests are successful and you've already spent too much effort. I think we should approve that PR. After merging, it would be easier to detect what can be done. Then, one can create a new PR to fix them. I try to follow but I've lost in changes. 🤕

I think we're waiting for Hacktober fest to open up

**Review by brylie — CHANGES_REQUESTED** [review]
Edit: removing the comment as I was mistaken.

File: project/core/urls.py
Comment (brylie) [inline project/core/urls.py:24]: This should be `threads_urls` now instead of `api`
```suggestion
from threads import urls as threads_urls
```

**Review by brylie — CHANGES_REQUESTED** [review]
The code, as it stands, fails to load after running `python manage.py collectstatic` with the following error:

> Exception Value: Missing staticfiles manifest entry for dependencies/materialize.min.css'

Please restore the `materialize.min.css` and `materialize.min.js` files and remove those as part of a separate pull request. Otherwise, we need to complete the remaining parts of #1051 in this pull request.

File: project/threads/templates/static/dependencies/materialize.min.css
Comment (brylie) [inline project/threads/templates/static/dependencies/materialize.min.css:1]: Let's restore this dependency and remove it in a separate pull request.

File: project/threads/templates/static/dependencies/materialize.min.js
Comment (brylie) [inline project/threads/templates/static/dependencies/materialize.min.js]: Let's restore this dependency and remove it in a separate pull request.

**Review by brylie — CHANGES_REQUESTED** [review]

**Review by brylie — CHANGES_REQUESTED** [review]
Overall, this looks great! As mentioned previously, please restore the Materialize CSS and JS files. We can remove those in a separate pull request so that this pull request can be merged.

We will merge this pull request after the files are restored and Hactoberfest begins. Thanks for your excellent work and patience on this task @AbhijithGanesh 😃

**Comment by AbhijithGanesh** [comment]
> Overall, this looks great! As mentioned previously, please restore the Materialize CSS and JS files. We can remove those in a separate pull request so that this pull request can be merged.
> 
> We will merge this pull request after the files are restored and Hactoberfest begins. Thanks for your excellent work and patience on this task @AbhijithGanesh 😃 

If I work on materialize css and create another PR , unmerged changes including this shall reflect creating unnecessary confusion. Would the be okay

**Comment by brylie** [comment]
> If I work on materialize CSS and create another PR, unmerged changes including this shall reflect creating unnecessary confusion. 

When creating a new branch, always start your branch from the `develop` branch. Then, make specific changes related to the relevant issue. In this manner, your pull request will contain only appropriate changes (and not unmerged changes from other branches.)

**Comment by AbhijithGanesh** [comment]
I understand brylie and I'll keep this in my mind

**Comment by AbhijithGanesh** [comment]
> > If I work on materialize CSS and create another PR, unmerged changes including this shall reflect creating unnecessary confusion. 
> 
> When creating a new branch, always start your branch from the `develop` branch. Then, make specific changes related to the relevant issue. In this manner, your pull request will contain only appropriate changes (and not unmerged changes from other branches.)

Can you please merge the PR since Hacktober has officially begun @brylie

**Comment by brylie** [comment]
@AbhijithGanesh, we couldn't merge the pull request without restoring the Materialize files since the project would not run correctly without them. So I have fixed the Materialize files and will now merge this pull request.

**Review by brylie — APPROVED** [review]
Looks good. Remember to pay close attention when changes are requested.

I have restored the Materialize files so that this project will run.

**Pull request merged**
